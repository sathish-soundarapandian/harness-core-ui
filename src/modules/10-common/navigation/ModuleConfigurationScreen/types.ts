/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { Entry } from 'contentful'
import type { IconName } from '@harness/icons'
import type { ModuleName } from 'framework/types/ModuleName'

export interface LottieContent {
  data: Record<any, any>
}

export interface CenterAlignedImageDesc {
  primaryText?: string
  secondoryText?: string
  imageUrl: string
}

export interface ModuleInfo {
  label: string
  icon: IconName
}

export interface ContentfulModuleResponse {
  identifier: ModuleName
  label: string
  data: Entry<CenterAlignedImageDesc | LottieContent>[]
}
