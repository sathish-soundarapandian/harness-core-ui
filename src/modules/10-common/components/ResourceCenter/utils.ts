/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import moment from 'moment'
import type { UserInfo } from 'services/cd-ng'

export const ON_PREM_RELEASE_NODE_LINK = 'https://ngdocs.harness.io/article/556wy85kbo-harness-on-prem-release-notes'
export const SAAS_RELEASE_NODE_LINK = 'https://ngdocs.harness.io/article/7zkchy5lhj-harness-saa-s-release-notes-2022'

export const HARNESS_SEARCH_LINK = 'https://harness.io/search/'
export const HARNESS_UNIVERISITY_LINK = 'https://university.harness.io/page/job-role-certifications'
export const HARNESS_COMMUNITY_LINK = 'https://community.harness.io/'
export const HARNESS_COMMUNITY_SLACK_LINK =
  'https://join.slack.com/t/harnesscommunity/shared_invite/zt-y4hdqh7p-RVuEQyIl5Hcx4Ck8VCvzBw'
export const HARNESS_DOCS_LINK = 'https://docs.harness.io/'
export const HARNESS_API_DOCS_LINK = 'https://apidocs.harness.io/'
export const HARNESS_TUTORIALS = 'https://developer.harness.io/tutorials/get-started'
export const SITE_STATUS_LINK = 'https://status.harness.io/'

export const timestamp = moment.now()
export const HARNESS_SUPPORT_LINK =
  '/sso.html?action=login&brand_id=114095000394&locale_id=1&return_to=https%3A%2F%2Fsupport.harness.io%2Fhc%2Fen-us%2Frequests&src=zendesk&timestamp=' +
  timestamp

export const WHATS_NEW_LINK = `https://docs.harness.io/article/ueeiji09do-what-s-new`
export const EARLY_ACCESS_LINK = `https://docs.harness.io/article/w4krvu96i3-early-access`

export const openZendeskSupport = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(HARNESS_SUPPORT_LINK)
}

export function getReleaseNodeLink(): string {
  switch (window.deploymentType) {
    case 'COMMUNITY':
    case 'ON_PREM': {
      return ON_PREM_RELEASE_NODE_LINK
    }
    default:
      return SAAS_RELEASE_NODE_LINK
  }
}
export const openFileATicket = (
  e: React.MouseEvent<Element, MouseEvent>,
  currentUserInfo: UserInfo,
  setShow?: React.Dispatch<React.SetStateAction<boolean>>
): void => {
  e.stopPropagation()
  e.preventDefault()
  window.Saber.do?.('set_options', {
    feedback_values: {
      Email: currentUserInfo.email // set default value for email field
    }
  })
  setShow?.(false)
  window.Saber.do?.('open')
}
export const openWhatsNew = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(WHATS_NEW_LINK)
}
export const openEarlyAccess = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(EARLY_ACCESS_LINK)
}
