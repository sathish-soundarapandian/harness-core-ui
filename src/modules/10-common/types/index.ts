import type { QlceViewTimeGroupType, QlceViewFilterInput, QlceViewFieldInputInput } from 'services/ce/services'

export interface TimeRangeFilterType {
  to: string
  from: string
}

export enum TimeRangeType {
  'LAST_7' = 'LAST_7',
  'LAST_30' = 'LAST_30'
}

export enum NodepoolTimeRangeType {
  'LAST_DAY' = 'LAST_DAY',
  'LAST_7' = 'LAST_7',
  'LAST_30' = 'LAST_30'
}

export type setFiltersFn = (newFilters: QlceViewFilterInput[]) => void
export type setAggregationFn = (newAgg: QlceViewTimeGroupType) => void
export type setGroupByFn = (groupBy: QlceViewFieldInputInput) => void
export type setTimeRangeFn = (timeRange: TimeRangeFilterType) => void
