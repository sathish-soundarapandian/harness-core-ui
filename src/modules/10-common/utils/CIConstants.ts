/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */
export const DEFAULT_ORG_ID = 'default'

export const DEFAULT_PROJECT_NAME = 'Default Project'

export const DEFAULT_PROJECT_ID = 'Default_Project'

export const DEFAULT_ORG_NAME = 'Default Organization'

export enum Status {
  SUCCESS = 'SUCCESS',
  ERROR = 'ERROR',
  FAILURE = 'FAILURE'
}
