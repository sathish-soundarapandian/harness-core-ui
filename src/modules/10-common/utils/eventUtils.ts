/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { SyntheticEvent } from 'react'

export function killEvent(e: React.MouseEvent<any> | SyntheticEvent<HTMLElement, Event> | undefined): void {
  // do not add preventDefault here, that works odd with checkbox selection
  e?.stopPropagation()
}
