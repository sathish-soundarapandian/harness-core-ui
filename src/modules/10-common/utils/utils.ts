/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { getMultiTypeFromValue, IconName, MultiSelectOption, MultiTypeInputType, SelectOption } from '@harness/uicore'
import { defaultTo, isUndefined } from 'lodash-es'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName, moduleToModuleNameMapping } from 'framework/types/ModuleName'
import { useLicenseStore } from 'framework/LicenseStore/LicenseStoreContext'
import { Editions } from '@common/constants/SubscriptionTypes'
import type { UserMetadataDTO } from 'services/cd-ng'
import type { Module } from '@common/interfaces/RouteInterfaces'
import { useModuleInfo } from '@common/hooks/useModuleInfo'

const PR_ENV_HOST_NAME = 'pr.harness.io'

interface SetPageNumberProps {
  setPage: (value: number) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
    case ModuleName.STO:
      return 'sto-color-filled'
    case ModuleName.CHAOS:
      return 'chaos-main'
    case ModuleName.ET:
      return 'cet'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}

export const formatCount = (num: number): string => {
  const min = 1e3
  if (num >= min) {
    const units = ['k', 'M', 'B', 'T']
    const order = Math.floor(Math.log(num) / Math.log(1000))
    const finalNum = Math.round(num / 1000 ** order)
    if (finalNum === min && order < units.length) {
      return 1 + units[order]
    }
    return finalNum + units[order - 1]
  }
  return num.toLocaleString()
}

export function useGetCommunity(): boolean {
  const { licenseInformation } = useLicenseStore()
  return licenseInformation?.['CD']?.edition === Editions.COMMUNITY
}

export const isOnPrem = (): boolean => window.deploymentType === 'ON_PREM'

export const focusOnNode = (node: HTMLElement): void => {
  const oldTabIndex = node.tabIndex
  node.tabIndex = -1
  node.focus()
  node.tabIndex = oldTabIndex
}

const HOTJAR_SUPPRESSION_ATTR = 'data-hj-suppress'

// Utility to add `data-hj-suppress` into a collection of elements to
// suppress data from HotJar recording
// @see https://bit.ly/3rCgpOY
export const suppressHotJarRecording = (elements: Element[] | null | undefined): void => {
  if (window.hj) {
    elements?.forEach?.((e: Element) => {
      if (!e.hasAttribute(HOTJAR_SUPPRESSION_ATTR)) {
        e.setAttribute(HOTJAR_SUPPRESSION_ATTR, 'true')
      }
    })
  }
}

// Utility to generate { 'data-hj-suppress': true } attribute if HotJar is available
export const addHotJarSuppressionAttribute = (): { [HOTJAR_SUPPRESSION_ATTR]: boolean } | undefined =>
  window.hj ? { [HOTJAR_SUPPRESSION_ATTR]: true } : undefined

// Utility to check if environment is a PR environment
export const isPR = (): boolean => {
  return location.hostname?.includes(PR_ENV_HOST_NAME)
}

// Utility to check if environment is a local develop environment
export const isLocalHost = (): boolean => {
  return location.hostname === 'localhost' || location.hostname === '127.0.0.1'
}

export const getPREnvNameFromURL = (url: string): string => {
  if (!url) {
    return ''
  }
  return isPR() ? url.split(PR_ENV_HOST_NAME)?.[1]?.split('/')?.[1] : ''
}

export function isMultiTypeFixed(type: MultiTypeInputType): boolean {
  return type === MultiTypeInputType.FIXED
}

export function isValueFixed(
  value?: boolean | string | number | SelectOption | string[] | MultiSelectOption[]
): boolean {
  const type = getMultiTypeFromValue(value)

  return isMultiTypeFixed(type)
}

export function isMultiTypeRuntime(type: MultiTypeInputType): boolean {
  return [MultiTypeInputType.EXECUTION_TIME, MultiTypeInputType.RUNTIME].includes(type)
}

export function isValueRuntimeInput(
  value?: boolean | string | number | SelectOption | string[] | MultiSelectOption[]
): boolean {
  const type = getMultiTypeFromValue(value)

  return isMultiTypeRuntime(type)
}

export function isMultiTypeExpression(type: MultiTypeInputType): boolean {
  return type === MultiTypeInputType.EXPRESSION
}

export function getLastURLPathParam(URL: string): string {
  const splitArray = URL.split('/')
  return splitArray.length > 0 ? splitArray[splitArray.length - 1]?.split('-')?.join('') : ''
}

export function isValueExpression(
  value?: boolean | string | number | SelectOption | string[] | MultiSelectOption[]
): boolean {
  const type = getMultiTypeFromValue(value)

  return isMultiTypeExpression(type)
}

export const getUserName = (user: UserMetadataDTO): string => {
  return defaultTo(user.name, user.email)
}

export const REFERER_URL = 'refererURL'

export const getSavedRefererURL = (): string => localStorage.getItem(REFERER_URL) || ''
export const getGaClientID = (): string => {
  try {
    return defaultTo(
      document?.cookie
        ?.split('; ')
        ?.find((key: string) => key.includes('_ga='))
        ?.split('.')
        ?.slice(2)
        ?.join('.'),
      ''
    )
  } catch (e) {
    return ''
  }
}

export const isWindowsOS = (): boolean => navigator?.appVersion?.indexOf('Win') != -1

export const toBase64 = (file: File): Promise<string> =>
  new Promise((resolve, reject) => {
    const reader = new FileReader()
    reader.onloadend = () => resolve((reader.result || '').toString())
    reader.onerror = reject
    reader.readAsDataURL(file)
  })

export const getImageDimensions = (url: string): Promise<{ width: number; height: number }> => {
  return new Promise((resolve, reject) => {
    const img = document.createElement('img')

    img.onload = () => {
      const width = img.naturalWidth
      const height = img.naturalHeight

      resolve({ width, height })
    }
    img.onerror = reject
    img.src = url
  })
}

export interface NumberFormatterOptions {
  truncate?: boolean
}

export const numberFormatter: (value?: number, options?: NumberFormatterOptions) => string = (
  value?: number,
  options = { truncate: true }
) => {
  // istanbul ignore if
  if (value === undefined) {
    return ''
  }
  const truncateOptions = [
    { value: 1000000, suffix: 'm' },
    { value: 1000, suffix: 'k' }
  ]
  if (options.truncate) {
    for (const truncateOption of truncateOptions) {
      if (value >= truncateOption.value) {
        const truncatedValue = value / truncateOption.value
        // istanbul ignore if
        if (truncatedValue % 1 !== 0) {
          return `${truncatedValue.toFixed(1)}${truncateOption.suffix}`
        }
        return `${truncatedValue}${truncateOption.suffix}`
      }
    }
  }
  return `${getFixed(value)}`
}

export const getFixed = (value: number, places = 1): number => {
  if (value % 1 === 0) {
    return value
  }
  return parseFloat(value.toFixed(places))
}

export const getRefFromIdentifier = (
  identifier: string,
  orgIdentifier?: string,
  projectIdentifier?: string
): string => {
  if (projectIdentifier) {
    return identifier
  } else if (!orgIdentifier) {
    return `account.${identifier}`
  }
  return `org.${identifier}`
}

export const getIdentifierFromScopedRef = (entity: string): string => {
  return entity.indexOf('.') < 0 ? entity : entity.split('.')[1]
}

export const countAllKeysInObject = (obj: Record<string, any>): number =>
  Object.keys(obj).reduce((acc, curr) => {
    if (typeof obj[curr] === 'object' && !Array.isArray(obj[curr]) && obj[curr] !== null) {
      return 1 + acc + countAllKeysInObject(obj[curr])
    } else {
      return 1 + acc
    }
  }, 0)

export const isSimplifiedYAMLEnabled = (module?: Module, isSimpliedYAMLFFEnabled?: boolean): boolean => {
  return isSimpliedYAMLFFEnabled
    ? module?.valueOf().toLowerCase() === moduleToModuleNameMapping.ci.toLowerCase() ||
        module?.valueOf().toLowerCase() === moduleToModuleNameMapping.iacm.toLowerCase()
    : false
}

export function useGetCDFree(): boolean {
  const { licenseInformation } = useLicenseStore()
  return licenseInformation?.['CD']?.edition === Editions.FREE
}

export const useGetFreeOrCommunityCD = (): boolean => {
  const { module } = useModuleInfo()
  const isCommunity = useGetCommunity()
  const isFree = useGetCDFree()
  // Visible at projects and cd all empty state pages
  return (module === 'cd' || isUndefined(module)) && (isFree || isCommunity)
}
