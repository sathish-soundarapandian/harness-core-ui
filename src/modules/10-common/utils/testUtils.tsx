/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { UseGetProps, UseGetReturn, RestfulProvider } from 'restful-react'
import { act, fireEvent, getByText, queryByAttribute, waitFor } from '@testing-library/react'
import { compile } from 'path-to-regexp'
import { createMemoryHistory } from 'history'
import { Router, Route, Switch, useLocation, useHistory } from 'react-router-dom'
import { ModalProvider } from '@harness/use-modal'
import qs from 'qs'
import { noop } from 'lodash-es'
import { enableMapSet } from 'immer'
import userEvent from '@testing-library/user-event'
import { AppStoreContext, AppStoreContextProps } from 'framework/AppStore/AppStoreContext'
import { LicenseStoreContext, LicenseStoreContextProps } from 'framework/LicenseStore/LicenseStoreContext'
import { LICENSE_STATE_VALUES } from 'framework/LicenseStore/licenseStoreUtil'
import { withAccountId, accountPathProps } from '@common/utils/routeUtils'
import type { Project } from 'services/cd-ng'
import { StringsContext, UseStringsReturn } from 'framework/strings'

import { FeaturesContext, FeaturesContextProps } from 'framework/featureStore/FeaturesContext'
import type { FeatureDetail, FeatureMetaData } from 'framework/featureStore/featureStoreUtil'
import type { FeatureIdentifier } from 'framework/featureStore/FeatureIdentifier'
import './testUtils.scss'
import { PermissionsContext, PermissionsContextProps } from 'framework/rbac/PermissionsContext'
import { Editions } from '@common/constants/SubscriptionTypes'
import type { FeatureFlag } from '@common/featureFlags'
import { PreferenceStoreContext } from 'framework/PreferenceStore/PreferenceStoreContext'
import {
  TemplateSelectorContext,
  TemplatesSelectorContextInterface
} from 'framework/Templates/TemplateSelectorContext/TemplateSelectorContext'

export type UseGetMockData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Required<
  UseGetProps<TData, TError, TQueryParams, TPathParams>
>['mock']

export interface UseGetMockDataWithMutateAndRefetch<T> extends UseGetMockData<T> {
  mutate: () => Record<string, unknown>
  refetch: () => Record<string, unknown>
  cancel?: () => Record<string, unknown>
}

export interface UseMutateMockData<TData, TRequestBody = unknown> {
  loading?: boolean
  mutate?: (data?: TRequestBody) => Promise<TData>
}

export type UseGetReturnData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Omit<
  UseGetReturn<TData, TError, TQueryParams, TPathParams>,
  'absolutePath' | 'cancel' | 'response'
>

export const findDialogContainer = (): HTMLElement | null => document.querySelector('.bp3-dialog')
export const findPopoverContainer = (): HTMLElement | null => document.querySelector('.bp3-popover-content')

export interface TestWrapperProps {
  path?: string
  pathParams?: Record<string, string | number>
  queryParams?: Record<string, unknown>
  defaultAppStoreValues?: Partial<AppStoreContextProps>
  defaultLicenseStoreValues?: Partial<LicenseStoreContextProps>
  defaultPermissionValues?: Partial<PermissionsContextProps>
  defaultTemplateSelectorValues?: Partial<TemplatesSelectorContextInterface>
  defaultFeaturesValues?: Partial<FeaturesContextProps>
  defaultFeatureFlagValues?: Partial<Record<FeatureFlag, boolean>>
  projects?: Project[]
  enableBrowserView?: boolean
  stringsData?: Record<string, string>
  getString?: UseStringsReturn['getString']
}

export const prependAccountPath = (path: string): string => withAccountId(() => path)(accountPathProps)

export const CurrentLocation = (): JSX.Element => {
  const location = useLocation()
  return (
    <div>
      <h1>Not Found</h1>
      <div data-testid="location">{`${location.pathname}${
        location.search ? `?${location.search.replace(/^\?/g, '')}` : ''
      }`}</div>
    </div>
  )
}

export interface BrowserViewProps {
  enable?: boolean
  children: React.ReactNode
}

export function BrowserView(props: BrowserViewProps): React.ReactElement {
  const { enable, children } = props
  const location = useLocation()
  const history = useHistory()

  if (!enable) {
    return <>{children}</>
  }

  function handlePathChange(e: React.ChangeEvent<HTMLInputElement>): void {
    history.replace(e.currentTarget.value)
  }

  const search = location.search ? `?${location.search.replace(/^\?/, '')}` : ''

  return (
    <div className="browser">
      <div className="browser-header">
        <input className="browser-path" value={location.pathname + search} onChange={handlePathChange} />
      </div>
      <div className="browser-content">{children}</div>
    </div>
  )
}

export const TestWrapper: React.FC<TestWrapperProps> = props => {
  enableMapSet()
  const {
    path = '/',
    pathParams = {},
    defaultAppStoreValues,
    queryParams = {},
    defaultLicenseStoreValues,
    defaultPermissionValues,
    defaultTemplateSelectorValues,
    defaultFeaturesValues,
    defaultFeatureFlagValues = {},
    stringsData = {},
    getString = (key: string) => key
  } = props

  const search = qs.stringify(queryParams, { addQueryPrefix: true })
  const routePath = compile(path)(pathParams) + search

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const history = React.useMemo(() => createMemoryHistory({ initialEntries: [routePath] }), [])

  const defaultReturn = {
    enabled: true
  }

  /** TODO: Try fixing this later. This is causing some tests to fail */
  // React.useEffect(() => {
  //   history.replace(compile(path)(pathParams) + qs.stringify(queryParams, { addQueryPrefix: true }))
  //   // eslint-disable-next-line react-hooks/exhaustive-deps
  // }, [path, pathParams, queryParams])

  return (
    <Router history={history}>
      <StringsContext.Provider value={{ data: stringsData as any, getString }}>
        <PreferenceStoreContext.Provider
          value={{
            set: noop,
            get: noop,
            clear: noop
          }}
        >
          <AppStoreContext.Provider
            value={{
              featureFlags: {
                FEATURE_ENFORCEMENT_ENABLED: true,
                ...defaultFeatureFlagValues
              },
              updateAppStore: () => void 0,
              currentUserInfo: { uuid: '' },
              ...defaultAppStoreValues
            }}
          >
            <LicenseStoreContext.Provider
              value={{
                versionMap: {},
                licenseInformation: {},
                CI_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                FF_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CCM_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CD_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CHAOS_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                STO_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CV_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CET_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                updateLicenseStore: () => void 0,
                ...defaultLicenseStoreValues
              }}
            >
              <PermissionsContext.Provider
                value={{
                  permissions: new Map<string, boolean>(),
                  requestPermission: () => void 0,
                  checkPermission: () => true,
                  cancelRequest: () => void 0,
                  ...defaultPermissionValues
                }}
              >
                <FeaturesContext.Provider
                  value={{
                    features: new Map<FeatureIdentifier, FeatureDetail>(),
                    featureMap: new Map<FeatureIdentifier, FeatureMetaData>(),
                    getEdition: () => {
                      return undefined
                    },
                    requestFeatures: () => void 0,
                    requestLimitFeature: () => void 0,
                    checkFeature: () => {
                      return defaultReturn
                    },
                    checkLimitFeature: () => {
                      return defaultReturn
                    },
                    getRestrictionType: () => {
                      return undefined
                    },
                    getHighestEdition: () => {
                      return Editions.FREE
                    },
                    ...defaultFeaturesValues
                  }}
                >
                  <ModalProvider>
                    <TemplateSelectorContext.Provider
                      value={{
                        state: {
                          isDrawerOpened: false,
                          selectorData: {
                            templateType: 'Step',
                            onSubmit: jest.fn(),
                            onCancel: jest.fn()
                          }
                        },
                        openTemplateSelector: jest.fn(),
                        closeTemplateSelector: jest.fn(),
                        ...defaultTemplateSelectorValues
                      }}
                    >
                      <RestfulProvider base="/">
                        <BrowserView enable={props.enableBrowserView}>
                          <Switch>
                            <Route exact path={path}>
                              {props.children}
                            </Route>
                            <Route>
                              <CurrentLocation />
                            </Route>
                          </Switch>
                        </BrowserView>
                      </RestfulProvider>
                    </TemplateSelectorContext.Provider>
                  </ModalProvider>
                </FeaturesContext.Provider>
              </PermissionsContext.Provider>
            </LicenseStoreContext.Provider>
          </AppStoreContext.Provider>
        </PreferenceStoreContext.Provider>
      </StringsContext.Provider>
    </Router>
  )
}

export const queryByNameAttribute = (name: string, container: HTMLElement): HTMLElement | null =>
  queryByAttribute('name', container, name)

export const doConfigureOptionsTesting = async (
  cogModal: HTMLElement,
  fieldElement: HTMLInputElement
): Promise<void> => {
  // Type regex and submit
  // check if field has desired value
  await waitFor(() => expect(getByText(cogModal, 'common.configureOptions.regex')).toBeInTheDocument())
  const regexRadio = getByText(cogModal, 'common.configureOptions.regex')
  userEvent.click(regexRadio)
  const regexTextArea = queryByAttribute('name', cogModal, 'regExValues') as HTMLInputElement
  act(() => {
    fireEvent.change(regexTextArea, { target: { value: '<+input>.includes(/test/)' } })
  })
  const cogSubmit = getByText(cogModal, 'submit')
  userEvent.click(cogSubmit)
  await waitFor(() => expect(fieldElement.value).toBe('<+input>.regex(<+input>.includes(/test/))'))
}
