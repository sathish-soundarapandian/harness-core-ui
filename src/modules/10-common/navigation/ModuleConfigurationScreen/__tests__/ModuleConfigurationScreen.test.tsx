/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { ModuleName } from 'framework/types/ModuleName'
import { usePreferenceStore } from 'framework/PreferenceStore/PreferenceStoreContext'
import ModulesConfigurationScreen from '../ModuleConfigurationScreen'
import type { ModuleCarouselProps } from '../ModuleDetailsSection/ModuleCarousel'
import type { ModuleSortableListProps } from '../ModuleSortableList/ModuleSortableList'

jest.mock('framework/PreferenceStore/PreferenceStoreContext')

jest.mock('../useGetContentfulModules', () => {
  return () => {
    return {
      contentfulModuleMap: {},
      loading: false
    }
  }
})

const setModuleConfigPreference = jest.fn()

jest.mock('../ModuleSortableList/ModuleSortableList', () => {
  return (props: ModuleSortableListProps) => {
    return (
      <>
        <div data-testId={'selected-modules-length'}>{props.selectedModules.length}</div>
      </>
    )
  }
})

jest.mock('../ModuleDetailsSection/ModuleCarousel', () => {
  return (props: ModuleCarouselProps) => {
    return <>{props.module ? <div data-testId={`test-active-module-${props.module}`}></div> : null}</>
  }
})

describe('Module Configuration screen', () => {
  beforeEach(() => {
    ;(usePreferenceStore as jest.Mock).mockImplementation(() => {
      return {
        setPreference: setModuleConfigPreference,
        preference: {
          orderedModules: [ModuleName.CD, ModuleName.CI],
          selectedModules: [ModuleName.CD]
        },
        clearPreference: jest.fn
      }
    })
  })
  test('should render correctly', () => {
    const { container, queryByText } = render(
      <TestWrapper path="/account/:accountId" pathParams={{ accountId: 'dummy' }}>
        <ModulesConfigurationScreen onClose={jest.fn} />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
    expect(queryByText('common.moduleConfig.selectModules')).toBeDefined()
    expect(queryByText('Module Sortable List')).toBeDefined()
    expect(queryByText('Module Carousel')).toBeDefined()
  })

  test('test when hide reordering is enabled', () => {
    const { queryByText } = render(
      <TestWrapper path="/account/:accountId" pathParams={{ accountId: 'dummy' }}>
        <ModulesConfigurationScreen hideReordering onClose={jest.fn} />
      </TestWrapper>
    )

    expect(queryByText('common.moduleConfig.selectModules')).toBeDefined()
    expect(queryByText('Module Sortable List')).toBeNull()
    expect(queryByText('Module Carousel')).toBeDefined()
  })

  test('test when hide header is enabled', () => {
    const { queryByText } = render(
      <TestWrapper path="/account/:accountId" pathParams={{ accountId: 'dummy' }}>
        <ModulesConfigurationScreen hideReordering hideHeader onClose={jest.fn} />
      </TestWrapper>
    )

    expect(queryByText('common.moduleConfig.selectModules')).toBeNull()
    expect(queryByText('Module Sortable List')).toBeNull()
    expect(queryByText('Module Carousel')).toBeDefined()
  })

  test('test with passing active module', () => {
    const { getByTestId } = render(
      <TestWrapper path="/account/:accountId" pathParams={{ accountId: 'dummy' }}>
        <ModulesConfigurationScreen activeModuleIndex={1} onClose={jest.fn} />
      </TestWrapper>
    )

    expect(getByTestId('test-active-module-CI')).toBeDefined()
  })

  test('test with no modules in preference store', async () => {
    ;(usePreferenceStore as jest.Mock).mockImplementation(() => {
      return {
        setPreference: setModuleConfigPreference,
        preference: {},
        clearPreference: jest.fn
      }
    })
    const { findByTestId } = render(
      <TestWrapper path="/account/:accountId" pathParams={{ accountId: 'dummy' }}>
        <ModulesConfigurationScreen activeModuleIndex={0} onClose={jest.fn} />
      </TestWrapper>
    )
    const selectedModulesLength = await findByTestId('selected-modules-length')

    expect(selectedModulesLength.innerHTML).toEqual('0')
  })
})
