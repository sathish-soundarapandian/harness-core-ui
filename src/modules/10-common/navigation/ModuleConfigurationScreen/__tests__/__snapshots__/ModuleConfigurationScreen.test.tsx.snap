// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module Configuration screen should render correctly 1`] = `
<div>
  <div
    class="Layout--vertical StyledProps--main container StyledProps--padding-left-xlarge"
  >
    <div
      class="StyledProps--font StyledProps--main header"
    >
      <span
        class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--margin-bottom-xsmall StyledProps--flex StyledProps--flex-justifyContent-flex-start"
      >
        <span
          class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--color StyledProps--color-white StyledProps--font-variation-h2"
        >
          <span>
            common.moduleConfig.selectModules
          </span>
        </span>
        <span
          class="StyledProps--font StyledProps--main blueText StyledProps--inline StyledProps--color StyledProps--color-primary5 StyledProps--margin-left-small StyledProps--margin-right-small"
        >
          <span>
            common.moduleConfig.your
          </span>
        </span>
        <span
          class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--color StyledProps--color-white StyledProps--font-variation-h2"
        >
          <span>
            common.moduleConfig.navigation
          </span>
        </span>
      </span>
      <p
        class="StyledProps--font StyledProps--main defaultSettingsTextContainer"
      >
        <span
          class="StyledProps--font StyledProps--main StyledProps--font-variation-small StyledProps--color StyledProps--color-grey200 StyledProps--inline"
        >
          (
          <span>
            common.moduleConfig.autoSaved
          </span>
          )
        </span>
      </p>
    </div>
    <div
      class="Layout--horizontal StyledProps--main body StyledProps--padding-bottom-huge StyledProps--padding-right-huge StyledProps--margin-bottom-xxxlarge"
    >
      <div
        class="StyledProps--font StyledProps--main sortableListContainer StyledProps--margin-right-xxlarge"
      >
        <div
          data-testid="selected-modules-length"
        >
          1
        </div>
      </div>
      <div
        class="StyledProps--font StyledProps--main flex1"
      >
        <div
          data-testid="test-active-module-CD"
        />
      </div>
    </div>
    <span
      class="bp3-icon bp3-icon-cross StyledProps--main crossIcon StyledProps--color StyledProps--color-white"
      icon="cross"
    >
      <svg
        data-icon="cross"
        height="18"
        viewBox="0 0 16 16"
        width="18"
      >
        <desc>
          cross
        </desc>
        <path
          d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </div>
</div>
`;
